### DISCUSSION TITLE
List of Office IDs?

### DISCUSSION THREAD

**Discussion by mousedownmike** [weather-gov/api#690] [Q&A] [answered]
The API supports getting details for a NWS office based on ID using `/offices/{officeId}`.  However, I can't seem to find a reliable source for the list of office IDs.  

1. Is this data available through the API other than in an error response (see below)?
2. If not, is the list of offices static enough to simply include in my application?

If I submit an invalid `officeId`, I do get three arrays of valid values that, when combined, result in the following 138 possible IDs:

```json
["NWS","AKQ","ALY","BGM","BOX","BTV","BUF","CAE","CAR","CHS","CLE","CTP","GSP","GYX","ILM","ILN","LWX","MHX","OKX","PBZ","PHI","RAH","RLX","RNK","ABQ","AMA","BMX","BRO","CRP","EPZ","EWX","FFC","FWD","HGX","HUN","JAN","JAX","KEY","LCH","LIX","LUB","LZK","MAF","MEG","MFL","MLB","MOB","MRX","OHX","OUN","SHV","SJT","SJU","TAE","TBW","TSA","ABR","APX","ARX","BIS","BOU","CYS","DDC","DLH","DMX","DTX","DVN","EAX","FGF","FSD","GID","GJT","GLD","GRB","GRR","ICT","ILX","IND","IWX","JKL","LBF","LMK","LOT","LSX","MKX","MPX","MQT","OAX","PAH","PUB","RIW","SGF","TOP","UNR","BOI","BYZ","EKA","FGZ","GGW","HNX","LKN","LOX","MFR","MSO","MTR","OTX","PDT","PIH","PQR","PSR","REV","SEW","SGX","SLC","STO","TFX","TWC","VEF","AER","AFC","AFG","AJK","ALU","GUM","HPA","HFO","PPG","STU","NH1","NH2","ONA","ONP","ARH","CRH","ERH","PRH","SRH","WRH"]
```

**Comment by mousedownmike** [comment]
I should note that I do see those same IDs enumerated in the OpenAPI specification. That just doesn't seem like a great way to programmatically get the list of IDs.  I was hoping for something similar to `/stations` or `/zones`.

**Reply by dwhitemv25** [reply]
Out of curiosity, why do you need this? There are no queries that use the WFO ID as a key (other than the `/offices` queries), and you can use `/points` to get the WFO for a given location.

**Reply by mousedownmike** [reply]
I'm not sure what a WFO is.  I'm looking for the list of NWS office IDs to get an area forecast discussion (AFD product).

**Reply by dwhitemv25** [reply]
WFO = Weather Forecast Office

You need to get _all_ the AFDs from _every_ WFO?

Edit: Sorry if you didn't like the question. I asked it because if you need to request the AFDs in bulk I know of a more efficient method.

**Comment by JJKraw** [comment, marked as answer]
I use

https://api.weather.gov/products/locations

Mostly null entries but I believe all of the offices are in there.

**Reply by JJKraw** [reply]
And as to "why?", I have a cron job that runs every couple of weeks that fetches the office detail for each office and caches it. That way I don't have to fetch the info in real time when I need it (to display info about the issuing office re: alerts, products, forecasts).

**Comment by codybrom** [comment]
I took the list at `https://api.weather.gov/products/locations` and checked every one of the 1657 items against the `https://api.weather.gov/offices/{officeId}` endpoint.

What I got was 127 valid options:
`[ "ABQ", "ABR", "AER", "AFC", "AFG", "AJK", "AKQ", "ALU", "ALY", "AMA", "APX", "ARX", "BGM", "BIS", "BMX", "BOI", "BOU", "BOX", "BRO", "BTV", "BUF", "BYZ", "CAE", "CAR", "CHS", "CLE", "CRH", "CRP", "CTP", "CYS", "DDC", "DLH", "DMX", "DTX", "DVN", "EAX", "EKA", "EPZ", "ERH", "EWX", "FFC", "FGF", "FGZ", "FSD", "FWD", "GGW", "GID", "GJT", "GLD", "GRB", "GRR", "GSP", "GUM", "GYX", "HFO", "HGX", "HNX", "HUN", "ICT", "ILM", "ILN", "ILX", "IND", "IWX", "JAN", "JAX", "JKL", "KEY", "LBF", "LCH", "LIX", "LKN", "LMK", "LOT", "LOX", "LSX", "LUB", "LWX", "LZK", "MAF", "MEG", "MFL", "MFR", "MHX", "MKX", "MLB", "MOB", "MPX", "MQT", "MRX", "MSO", "MTR", "OAX", "OHX", "OKX", "OTX", "OUN", "PAH", "PBZ", "PDT", "PHI", "PIH", "PQR", "PSR", "PUB", "RAH", "REV", "RIW", "RLX", "RNK", "SEW", "SGF", "SGX", "SHV", "SJT", "SJU", "SLC", "SRH", "STO", "TAE", "TBW", "TFX", "TOP", "TSA", "TWC", "UNR", "VEF"]`

This list is valid as of 2/15/2024 at 9:45 PM CST.

**Comment by greg-does-weather** [comment]
There is an open ticket in the API to add an endpoint for the list of offices. No ETA on that, just FYI that it's in the hopper. 🙂
